openmls/src/ciphersuite/mod.rs

```diff
@@ -546,63 +549,150 @@ impl Signature {
+            if scalar_length > P256_SCALAR_LENGTH + 1 {
+                return Err(SignatureError::DecodingError);
+            };
+
+            // If the scalar is 0x21 long, the first byte has to be 0x00,
+            // indicating that the following integer is unsigned. We can discard
+            // this byte safely. If it's not 0x00, the scalar is too large not
+            // thus not a valid point on the curve.
+            if scalar_length == P256_SCALAR_LENGTH + 1 {
+                if buffer.read_u8()? != 0x00 {
+                    return Err(SignatureError::DecodingError);
+                };
+                // Since we just read that byte, we decrease the length by 1.
+                scalar_length -= 1;
+            };
+
+            let mut scalar = vec![0; scalar_length];
+            buffer.read_exact(&mut scalar)?;
+
+            // The scalar should have the right length.
+            if scalar.len() != scalar_length {
+                return Err(SignatureError::DecodingError);
+            };
+
             // The verification algorithm expects the scalars to be 32 bytes
             // long, buffered with zeroes.
-            let mut padded_scalar = vec![0u8; SCALAR_SIZE - decoded_scalar_len];
-            padded_scalar.extend_from_slice(decoded_scalar.as_bytes());
+            let mut padded_scalar = vec![0u8; P256_SCALAR_LENGTH - scalar_length];
+            padded_scalar.extend_from_slice(&scalar);
+
             Ok(padded_scalar)
+        }
+
+        let mut signature_bytes = self.value.as_slice();
+
+        // Check header bytes:
+        // 1 byte SEQUENCE tag should be 0x30
+        let sequence_tag = signature_bytes.read_u8()?;
+        if sequence_tag != SEQUENCE_TAG {
+            return Err(SignatureError::DecodingError);
         };
 
-        let mut out = Vec::with_capacity(2 * SCALAR_SIZE + MAX_DER_OVERHEAD);
-        let (r, s) = der::Decoder::new(self.value.as_slice()).sequence(|decoder| {
-            let r_decoded = decode_scalar(decoder)?;
-            let s_decoded = decode_scalar(decoder)?;
-            Ok((r_decoded, s_decoded))
-        })?;
+        // At most 1 byte encoding the length of the scalars (short form DER
+        // length encoding). Length has to be encoded in the short form, as we
+        // expect the length not to exceed the maximum length of 70: Two times
+        // at most 32 (scalar value) + 1 byte integer tag + 1 byte length tag +
+        // at most 1 byte to indicating that the integer is unsigned.
+        let length = signature_bytes.read_u8()? as usize;
+        if length > 2 * (P256_SCALAR_LENGTH + 3) {
+            return Err(SignatureError::DecodingError);
+        }
+
+        // The remaining bytes should be equal to the encoded length.
+        if signature_bytes.len() != length {
+            return Err(SignatureError::DecodingError);
+        }
+
+        let r = decode_scalar(&mut signature_bytes)?;
+        let s = decode_scalar(signature_bytes)?;
```

Comment:
Why is this not `&mut`?

Reply:
Honestly, I'm not sure what is happening under the hood here exactly. I know that the first one has to be `mut` for `decode_scalar` to move the read pointer forward. For the second call I'm not sure.

Reply:
I now made it `mut` so that we can afterwards check that indeed everything has been read from the `signature_bytes` reader.